parsers/moose.c

```diff
@@ -0,0 +1,273 @@
+{
+	struct mooseSubparser *moose = (struct mooseSubparser *)perl;
+	moose->inPod = false;
+}
+
+static bool leaveMoose (const char *line CTAGS_ATTR_UNUSED,
+						const regexMatch *matches CTAGS_ATTR_UNUSED,
+						unsigned int count CTAGS_ATTR_UNUSED,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	tagEntryInfo *e = getEntryInCorkQueue (moose->classCork);
+	Assert (e);
+	e->extensionFields.endLine = getInputLineNumber ();
+
+	moose->classCork = CORK_NIL;
+	moose->notInMoose = true;
+	moose->packageCork = CORK_NIL;
+	return true;
+}
+
+static bool enterMoose(const char *line CTAGS_ATTR_UNUSED,
+						const regexMatch *matches CTAGS_ATTR_UNUSED,
+						unsigned int count CTAGS_ATTR_UNUSED,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (moose->packageCork == CORK_NIL)
+		return true;
+
+	moose->notInMoose = false;
+	tagEntryInfo *perl_e = getEntryInCorkQueue (moose->packageCork);
+	Assert (perl_e);
+
+	tagEntryInfo moose_e;
+	initTagEntry (&moose_e, perl_e->name, K_CLASS);
+	moose_e.lineNumber = perl_e->lineNumber;
+	moose_e.filePosition = perl_e->filePosition;
+	moose->classCork = makeTagEntry (&moose_e);
+
+	return true;
+}
+
+static bool extendClass(const char *line,
+						const regexMatch *matches,
+						unsigned int count,
+						void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (moose->classCork == CORK_NIL)
+		return true;
+
+	tagEntryInfo *e = getEntryInCorkQueue (moose->classCork);
+	Assert (e);
+	if (e->extensionFields.inheritance == NULL)
+		e->extensionFields.inheritance = eStrndup (line + matches[1].start,
+												   matches[1].length);
+
+	return true;
+}
+
+static bool findAttribute(const char *line,
+						  const regexMatch *matches,
+						  unsigned int count,
+						  void *data)
+{
+	struct mooseSubparser *moose = data;
+
+	if (moose->inPod)
+		return true;
+
+	if (count < 3)
+		return true;
+
+	tagEntryInfo e;
+	char *buf = eStrndup (line + matches[2].start, matches[2].length);
+
+	initTagEntry (&e, buf, K_ATTR);
+	if (moose->classCork != CORK_NIL)
+		e.extensionFields.scopeIndex = moose->classCork;
+	makeTagEntry (&e);
+
+	eFree (buf);
+
+	return true;
+}
+
+static void findMooseTags(void)
+{
+	scheduleRunningBaseparser (RUN_DEFAULT_SUBPARSERS);
+}
+
+static void initializeMooseParser (langType language)
+{
+	addLanguageCallbackRegex (language, "^ *use +Moose(:| *;)",
+							  "{exclusive}",
+							  enterMoose, NULL,
+							  &mooseSubparser);
+	addLanguageCallbackRegex (language, "^ *no +Moose *;",
+							  "{exclusive}",
+							  leaveMoose, &mooseSubparser.notInMoose,
+							  &mooseSubparser);
+	addLanguageCallbackRegex (language, "^ *extends *'([^']+)';",
+							  "{exclusive}",
+							  extendClass, &mooseSubparser.notInMoose,
+							  &mooseSubparser);
+	addLanguageCallbackRegex (language, "^ *(has|after) *'([^']+)'",
```

Comment:
If you're going to do `after` you may as well do `around` and `before` just to finish them off. They all fit into a class of Aspect Oriented Programming modifiers that wrap methods provided by the parent class.

Reply:
As another special note, it seems from this regex that the `'` is required but in perl, they're not required (and they can and ofter are omitted entirely). As an added complexity they can be `"` (double quotes) too.

Reply:
This seems to only cover the `has 'literal1' =>`, but there is another form the `has [qw/literal1 literal2/]` form which is a shorthand to define many attributes with the same property.

Reply:
I see. I will add more commits for the comments.